### ISSUE TITLE
Within-chain parallelization via "reduce_sum"

### ISSUE THREAD

**Issue by paul-buerkner** [paul-buerkner/brms#892]
See the blog post of Sebastian Weber (@wds15): https://statmodeling.stat.columbia.edu/2020/05/05/easy-within-chain-parallelisation-in-stan/

**Comment by wds15** [comment]
Great to see this here.

It will likely take a moment until it lands in brms given that RStan is still at 2.19. Should we maybe prepare in the meantime a short vignette which shows with an example how the Stan code from the current brms needs to be modified in order to run with CmdStan? Is that an option? As alternative a wiki page maybe?

**Comment by paul-buerkner** [comment]
Would issue #891 help with this? :-)

**Comment by wds15** [comment]
Obviously yes.

**Comment by mike-lawrence** [comment]
Did the tutorial/wiki that @wds15 mentioned ever get created?

**Comment by paul-buerkner** [comment]
Inspired by @wds15 talk at StanCon, I started thinking about `reduce_sum` in brms a little more and I would like your (read: anybody's) input.

1) I am asking myself is over what the automatic slicing should be done. In a standard brms model, we could slice over `y` (the response variable) or over `mu` (the mean predictor term. A standard likelihood of brms could look as follows:

```
vector[N] y;
vector[N] mu;
real sigma;
target += normal_lpdf(y | mu, sigma);
```

As far as I understand, slicing over `mu` would be computationally prefferable as `y` can be copied by reference even if passed to `reduce_sum` via an additional argument, while `mu` would be passed by value if it was an additional argument.

**Comment by wds15** [comment]
1. Yep, slice over mu, not over y is much better.

2. It would still be worthwhile if the computational cost of the likelihood is very large compared to compiling mu serially. This is probably a good starting point. Still, is it thinkable to write the entire model block out as a function and just pass everything into it? 

I can't offer any benchmarks on this and I would suggest to just find out... I am looking forward to these results (speak: results someone else worked out). It's all about Amdahl's law as I explained...

**Comment by paul-buerkner** [comment]
Thanks! Let's aim high and parallelize the whole model block. Take, for example, the model block of a simple varying intercept-slope model, with priors removed (we don't want to include priors in reduce_sum, do we?):

```
model {
  // initialize linear predictor term
  vector[N] mu = Intercept + Xc * b;
  for (n in 1:N) {
    // add more terms to the linear predictor
    mu[n] += r_1_1[J_1[n]] * Z_1_1[n] + r_1_2[J_1[n]] * Z_1_2[n];
  }
  // likelihood including all constants
  if (!prior_only) {
    target += normal_lpdf(Y | mu, sigma);
  }
}
```

Here is how it would look like

```
functions {
  real partial_log_lik(real[] Y, int start, int end, real Intercept, 
                       matrix fXc, vector b, int[] fJ_1, vector fZ_1_1, 
                       vector fZ_1_2, vector r_1_1, vector r_1_2,
                       real sigma) {
    int N = end - start + 1;
    matrix[N, cols(fXc)] Xc = fXc[start:end];
    int J_1[N] = fJ_1[start:end];
    vector[N] Z_1_1 = fZ_1_1[start:end];
    vector[N] Z_1_2 = fZ_1_2[start:end];
    real ptarget = 0;
    // initialize linear predictor term
    vector[N] mu = Intercept + Xc * b;
    for (n in 1:N) {
      // add more terms to the linear predictor
      mu[n] += r_1_1[J_1[n]] * Z_1_1[n] + r_1_2[J_1[n]] * Z_1_2[n];
    }
    ptarget += normal_lpdf(Y | mu, sigma);
    return ptarget;
  }
}
model {
  // likelihood including all constants
  if (!prior_only) {
    target += reduce_sum(partial_log_lik, Y, 1, ...)
  }
}
```

where `...` are all the additional arguments passed to `partial_log_lik`.
I do a lot of housekeeping at the start of the partial_log_lik where I subset the original vectors etc. so that they become the partial vectors etc that we need. The full vectors to be subsetted get the prefix `f` so that the subsetted vectors have the original names. That way only the definition header of `partial_log_lik` is different from the original model block's code, which should make the required changes to brms' Stan code generation much less cumbersome.

Above, I use `Y` as the first argument not because it is necessarily the best to reduce over but because it is the only thing that is always available no matter how the model looks like.

@wds15 What do you think of this approach? Am I missing something important in the spec?

**Comment by wds15** [comment]
This starts to look very good! My thoughts on this:

- we first should try this out without all the code gen to see if this really gives us the cool effect we want. I would manually rewrite a generated brms model.
- the line `matrix[N, cols(fXc)] Xc = fXc[start:end];` is very expensive for the AD tape. The Xc will be declared as a parameter and add significantly to the AD cost. To avoid this, you would have to avoid declaring `Xc`, but rather write the expression with it directly... so replace `vector[N] mu = Intercept + Xc * b;` with `vector[N] mu = Intercept +  fXc[start:end] * b;`
- yes, priors should not go into the partial sum function.

**Comment by paul-buerkner** [comment]
I will write a brms branch where I support just a small subset of brms models in the above described way. That should take just a few hours but would enable us to rapidly test a lot of cases (instead of doing all the hand writing). 

With regard to `Xc`, can you explain why this specification is a problem for the AD tape while the same for, say, `Z_1_1` is not? Also, your suggestions would kill the principle of leaving the fundamental model block unchanged and just adding a definition header, so I would like to avoid it if possible.

**Comment by wds15** [comment]
The `Z_*` already come in as parameters and it looks to me as if doing this re-indexing (basically this is what it is) makes it easier for you to work within code-gen of brms.... but any real (real, vector, matrix) in a user functions is always instantiated as var (even if you assign data to it). At that point `Xc` is written to the AD tape. If you instead do what I suggested, then `Xc` is never written to the AD tape, since you use it in a matrix-vector multiplication. Doing so keeps the AD tape clean from the matrix and does a very much optimised matrix-vector product where only terms of the size of the vector end up on the AD tape.

However, if it is really hard to integrate that in your code gen, then just keep it as is for now (there is hope that the Stan compiler gets smarter and the problem goes away automatically).

**Comment by paul-buerkner** [comment]
I see. So, in brms, `Z_*` is data as well so I suspect the same would apply to those variables as well, right? That is, ideally don't redefine them but use them directly. The problem with this would be that `Z_1_1[start:end]` would be repeated in every evaluation of the loop, which is probably not what we want either... 

I will go ahead and do the header-only-change for now, even if inefficient in some cases so that we have a version to try out.

**Comment by wds15** [comment]
Oh... if that's also data... then yes, avoid the temporary re-assignemnts if you can.

So better use instead of
`Z_1_1[n]`
do this:
`fZ_1_1[start + n - 1]`

Then you keep this thing as data.

**Comment by paul-buerkner** [comment]
This is going to be a code-generation nightmare :-D Let's see if I can find a principled way to enable this indexing on the fly.

**Comment by wds15** [comment]
I usually write partial sum functions with a loop over the slice, but with two indices. One spanning `n=1...N` for `N` being the slice size and another one for the global context `n_overall=n + start - 1`. With these two defined it's often easy to write down what you want... but I understand that this may be harder for your brms code gen things.

**Comment by paul-buerkner** [comment]
Thanks. I will see what I can make possible in a first step.

**Comment by wds15** [comment]
Maybe start this by taking care of Xc... which is of the order of N * number of regressors.... but Z is just of the size of N.

**Comment by paul-buerkner** [comment]
Yeah, but in order to make that scale to all the complexity of brms models, we should consistently use one or the other approach all the way through and not mix things up. I will play around with it and report back once there is something that works (ish).

**Comment by paul-buerkner** [comment]
Ok, one follow up question, given that the header-change approach would be **so much** simpler: Do you know of any plans to make the Stan compiler smarter in definining variables as data (when appropriate) in user-defined functions?

**Comment by wds15** [comment]
I don't know... with some luck this is even already available with the recently release stanc3 optimisations which can be turned on. I have not been following this that close enough to know if it is there or is still in the making.

I would have hoped that a simple replace - text approach would work to start with at least.

**Comment by rok-cesnovar** [comment]
> Do you know of any plans to make the Stan compiler smarter in definining variables as data (when appropriate) in user-defined functions?

This is in 2.24 already. If you specify --O (not --o) it will optimize this case. Optimizations are currently experimental, but at least these optimizations work as they should (at least as far as we tested them off course).

**Comment by rok-cesnovar** [comment]
So 
```stan
functions {
real foo(data matrix X, vector paramv) {
   matrix[2,2] Xs = X[1:2];
   return sum( Xs * paramv);
}
}
```
without optimization:
```cpp
template <typename T1__>
stan::promote_args_t<T1__>
foo(const Eigen::Matrix<double, -1, -1>& X,
    const Eigen::Matrix<T1__, -1, 1>& paramv, std::ostream* pstream__) {
  using local_scalar_t__ = stan::promote_args_t<T1__>;
  const static bool propto__ = true;
  (void) propto__;
  local_scalar_t__ DUMMY_VAR__(std::numeric_limits<double>::quiet_NaN());
  (void) DUMMY_VAR__;  // suppress unused var warning
  
  try {
    Eigen::Matrix<local_scalar_t__, -1, -1> Xs;
    Xs = Eigen::Matrix<local_scalar_t__, -1, -1>(2, 2);
    stan::math::fill(Xs, DUMMY_VAR__);
    
    current_statement__ = 1;
    assign(Xs, nil_index_list(),
      rvalue(X, cons_list(index_min_max(1, 2), nil_index_list()), "X"),
      "assigning variable Xs");
    current_statement__ = 2;
    return sum(multiply(Xs, paramv));
  } catch (const std::exception& e) {
    stan::lang::rethrow_located(e, locations_array__[current_statement__]);
      // Next line prevents compiler griping about no return
      throw std::runtime_error("*** IF YOU SEE THIS, PLEASE REPORT A BUG ***"); 
  }
  
}
```
with optimization

```cpp
template <typename T1__>
stan::promote_args_t<T1__>
foo(const Eigen::Matrix<double, -1, -1>& X,
    const Eigen::Matrix<T1__, -1, 1>& paramv, std::ostream* pstream__) {
  using local_scalar_t__ = stan::promote_args_t<T1__>;
  const static bool propto__ = true;
  (void) propto__;
  local_scalar_t__ DUMMY_VAR__(std::numeric_limits<double>::quiet_NaN());
  (void) DUMMY_VAR__;  // suppress unused var warning
  
  try {
    Eigen::Matrix<double, -1, -1> lcm_sym2__;
    double lcm_sym1__;
    {
      Eigen::Matrix<double, -1, -1> Xs;
      Xs = Eigen::Matrix<double, -1, -1>(2, 2);
      stan::math::fill(Xs, std::numeric_limits<double>::quiet_NaN());
      
      assign(lcm_sym2__, nil_index_list(),
        rvalue(X, cons_list(index_min_max(1, 2), nil_index_list()), "X"),
        "assigning variable lcm_sym2__");
      current_statement__ = 2;
      return sum(multiply(lcm_sym2__, paramv));
    }
  } catch (const std::exception& e) {
    stan::lang::rethrow_located(e, locations_array__[current_statement__]);
      // Next line prevents compiler griping about no return
      throw std::runtime_error("*** IF YOU SEE THIS, PLEASE REPORT A BUG ***"); 
  }
  
}
```

Note that Xs is a matrix of doubles which is a huge win in this case.

**Comment by wds15** [comment]
Thanks @rok-cesnovar ... but a requirement is that you declare things as

```stan
real foo(data matrix X, vector paramv) {
   data matrix[2,2] Xs = X[1:2];
   return sum( Xs * paramv);
}
```

So the `X` must be declared as `data matrix` and one also can only call it with data at that point, of course.

Is that an option for you @paul-buerkner ?

(BTW, this is super cool to know... I have been waiting for this feature since ages)

**Comment by rok-cesnovar** [comment]
Yes, you need the data qualifier.

Otherwise stanc3 cant infer the type. In theory it probably could check with what inputs it is used and create extra signatures. But not atm.

**Comment by paul-buerkner** [comment]
Nice! So I just need to put the `data` qualifier before the variable type and things should be fine? That would be awesome!

**Comment by rok-cesnovar** [comment]
Yes + call the cmdstan_model with an argument. I will post an example call when I am back at my laptop.

**Comment by paul-buerkner** [comment]
Great! @wds15 After some work today, I got a simple version working already, which used the subsetting on the fly approach, you advocated before knowing of the new stanc optimization. I will post the version tomorrow or so when I am convinced of it. I am no longer sure which of the two discussed versions would be easier to implement and maintain. I will play around with it more.

**Comment by wds15** [comment]
If you get it to work with subsetting as suggested first, then this will be more flexible... since the `data` declaration in the arguments makes any function less flexible which could limit things wrt to missing data stuff, for example.

**Comment by rok-cesnovar** [comment]
Forgot about this, sorry..

```
library(cmdstanr)

file_path <- file.path(cmdstan_path(), "examples", "bernoulli", "bernoulli.stan")
mod <- cmdstan_model(file_path, stanc_options = list(O=TRUE))
```
You can run cmdstan_model with quiet = FALSE to double check. 

You should see something like: 
```
--- Translating Stan model to C++ code ---
bin/stanc --O --name='bernoulli_model' --o=/tmp/Rtmp2EuAeC/model-1c3a2fc5c314.hpp /tmp/Rtmp2EuAeC/model-1c3a2fc5c314.stan
```

**Comment by paul-buerkner** [comment]
brms now supports threading via `reduce_sum` for most models. However, this is super experimental and a lot of cleaning will be required. Here is an example for a very simple model:

```R
data("sleepstudy", package = "lme4")
```

```R
make_stancode(Reaction ~ Days + (Days | Reaction), sleepstudy,
              family = gaussian(), threads = 2)
```

```Stan
// generated with brms 2.13.10
functions {
  /* integer sequence of values
   * Args: 
   *   start: starting integer
   *   end: ending integer
   * Returns: 
   *   an integer sequence from start to end
   */ 
  int[] sequence(int start, int end) { 
    int seq[end - start + 1];
    for (n in 1:num_elements(seq)) {
      seq[n] = n + start - 1;
    }
    return seq; 
  } 
  // compute partial sums of the log-likelihood
  real partial_log_lik(int[] seq, int start, int end, vector Y, matrix Xc, vector b, real Intercept, real sigma, int[] J_1, vector Z_1_1, vector Z_1_2, vector r_1_1, vector r_1_2) {
    real ptarget = 0;
    int N = end - start + 1;
    // initialize linear predictor term
    vector[N] mu = Intercept + Xc[start:end] * b;
    for (n in 1:N) {
      // add more terms to the linear predictor
      int nn = n + start - 1;
      mu[n] += r_1_1[J_1[nn]] * Z_1_1[nn] + r_1_2[J_1[nn]] * Z_1_2[nn];
    }
      ptarget += normal_lpdf(Y[start:end] | mu, sigma);
    return ptarget;
  }
}
data {
  int<lower=1> N;  // total number of observations
  vector[N] Y;  // response variable
  int<lower=1> K;  // number of population-level effects
  matrix[N, K] X;  // population-level design matrix
  // data for group-level effects of ID 1
  int<lower=1> N_1;  // number of grouping levels
  int<lower=1> M_1;  // number of coefficients per level
  int<lower=1> J_1[N];  // grouping indicator per observation
  // group-level predictor values
  vector[N] Z_1_1;
  vector[N] Z_1_2;
  int<lower=1> NC_1;  // number of group-level correlations
  int prior_only;  // should the likelihood be ignored?
}
transformed data {
  int Kc = K - 1;
  matrix[N, Kc] Xc;  // centered version of X without an intercept
  vector[Kc] means_X;  // column means of X before centering
  for (i in 2:K) {
    means_X[i - 1] = mean(X[, i]);
    Xc[, i - 1] = X[, i] - means_X[i - 1];
  }
}
parameters {
  vector[Kc] b;  // population-level effects
  real Intercept;  // temporary intercept for centered predictors
  real<lower=0> sigma;  // residual SD
  vector<lower=0>[M_1] sd_1;  // group-level standard deviations
  matrix[M_1, N_1] z_1;  // standardized group-level effects
  cholesky_factor_corr[M_1] L_1;  // cholesky factor of correlation matrix
}
transformed parameters {
  matrix[N_1, M_1] r_1;  // actual group-level effects
  // using vectors speeds up indexing in loops
  vector[N_1] r_1_1;
  vector[N_1] r_1_2;
  // compute actual group-level effects
  r_1 = (diag_pre_multiply(sd_1, L_1) * z_1)';
  r_1_1 = r_1[, 1];
  r_1_2 = r_1[, 2];
}
model {
  // likelihood including all constants
  if (!prior_only) {
    int seq[N] = sequence(1, N);
    target += reduce_sum(partial_log_lik, seq, 1, Y, Xc, b, Intercept, sigma, J_1, Z_1_1, Z_1_2, r_1_1, r_1_2);
  }
  // priors including all constants
  target += student_t_lpdf(Intercept | 3, 288.7, 59.3);
  target += student_t_lpdf(sigma | 3, 0, 59.3)
    - 1 * student_t_lccdf(0 | 3, 0, 59.3);
  target += student_t_lpdf(sd_1 | 3, 0, 59.3)
    - 2 * student_t_lccdf(0 | 3, 0, 59.3);
  target += std_normal_lpdf(to_vector(z_1));
  target += lkj_corr_cholesky_lpdf(L_1 | 1);
}
generated quantities {
  // actual population-level intercept
  real b_Intercept = Intercept - dot_product(means_X, b);
  // compute group-level correlations
  corr_matrix[M_1] Cor_1 = multiply_lower_tri_self_transpose(L_1);
  vector<lower=-1,upper=1>[NC_1] cor_1;
  // extract upper diagonal of correlation matrix
  for (k in 1:M_1) {
    for (j in 1:(k - 1)) {
      cor_1[choose(k - 1, 2) + j] = Cor_1[j, k];
    }
  }
}
```

```R
fit1 <- brm(Reaction ~ Days + (Days | Reaction), sleepstudy,
            family = gaussian(), threads = 2, 
            chains = 1, backend = "cmdstanr")
```

On my windows machine threading slows down sampling drastically which may be because of windows, the simple model, or my implementation of `reduce_sum`. @wds15 Would you mind running some benchmarks on the brms implementation so that we can get an understanding how well we are doing at the moment?

**Comment by wds15** [comment]
Hmm... I did run a poisson example and I also see unreasonable slow downs. I need to debug this for a moment.

**Comment by paul-buerkner** [comment]
Thank you for looking into it!

**Comment by wds15** [comment]
Interesting... so I looked at the Poisson example `epilepsy`:

```R
fit1 <- brm(count ~ zAge + zBase * Trt + (1|patient),
            data = epilepsy, family = poisson(),
            threads = 1,
            chains = 1, backend = "cmdstanr")
```

and single thread runtime is 3.8s while the plain `reduce_sum` runtime was a whopping 30s. I tried a few things, but what made it finally work is to simply introduce a `grainsize` which I increased from the fixed value of 1 to 60 or so. Then the performance was ok (4.9s with 1 core / ~3s with 4 cores), but the example is still too small for real benefits (but at least I was able to beat single thread runtime).

So we need the grainsize... and it needs to be picked by the user.

... looking at this more closely, I think that what would really help is to slice not over the data items 1-N, but instead slice over the patients. To do that one would probably have to resort the data or create some clever index maps - which I am not sure if this is something which can be done in brms. If it's a possibility, then we should try that. The intuition behind that is that doing so minimizes unnecessary copies of the random effects (in fact, you would slice over the random effect) and also the work done per unit increases which makes very likely setting `grainsize` less important. Should this be an option, then I would just make it a convention that the random effect with the largest number of levels is picked to slice over (that's my guess which we should test a bit).

BTW... not for now... but it would be nice to let users choose between `reduce_sum` and `reduce_sum_static`. The static version is deterministic and setting `grainsize` well is a must, while `reduce_sum` should work ok with mediocre set `grainsizes` and adapt better to more/less CPU cores.

Overall this looks super promising!

**Comment by paul-buerkner** [comment]
Thanks you for looking into that! So `grainsize` it is then. How do we choose it appropriately? Or rather, how do we advice users to choose it?

Slicing over random effects is not an option for brms (in my current opinion), because in general, we not only have one but multiple (nested or crossed) effects and we can obviously slice only over one. Adding even more structure, such a splines or other special terms, and we won't be able to slice appropriately over any partical kind of parameter vector. And even if we would, we would only slice over one but not over the several others . There will be stuff copied over no matter what we do. I understand that this is something where we gain speed for some simple models but I don't see this being an option for brms unfortunately. 

I am happy to also support `reduce_sum_static`. Is there any difference other than the grainsize? We also need to think of a user-interface that allows to intuitively specify everything relevant in one place, that is, the number of `threads`, the `grainsize`, and the `static` option. I am thinking of an argument call `threading` (or so) and a corresponding function which takes and validates all the arguments to fully handle threading.

**Comment by wds15** [comment]
The random effects of the model usually groups the data into larger chunks. When we slice over these, we gain a few things here:

- choosing `grainsize` should be far less important as more work is done per unit
- we will not be in-efficient in terms of copying things as the sliced over random effect only is doubled

Maybe I will try this out on the epilepsy example to have a test drive of it.

In any case: We can add this additional bit later once convinced of the merits of it. What is there already is very useful as it looks to me.

`reduce_sum_static` will always form the exact same expression graph such that always the same partial sums are formed and summed together which makes everything exactly reproducible. The chunk size will always be close to, but never larger than `grainsize`.

`reduce_sum` uses `grainsize` just as a rough order for the chunk size and it tries to adapt things to the given CPUs and to the current load to make best use of the available resources (but the result is non deterministic).

Choosing grainsize... the Intel TBB answer: https://software.intel.com/en-us/node/506060 ... practically it's hard for me to give a definite answer... maybe start with # of data rows / ( 2 * maximal # of cores being used )... but don't nail me on that; it's very dependent on your problem & CPU you use.

**Comment by paul-buerkner** [comment]
Thanks for the additional information! This is really helpful. 

With regard to the slicing by random effects, I don't doubt the merits of this (but would still love to see some benchmarks). However, I don't see at this point how I can make this possible without a huge amount of special case coding that will haunt me for years when having to maintain that.  If someone comes up with an idea how this can be represented without substantially breaking the existing structure of the code generation, I will consider it of course.

**Comment by jgabry** [comment]
This is really cool. Thank you both for working on it! I just wanted to chime in with one suggestion about being cautious when advertising this feature: 

Basically, unless we have really solid advice on getting the settings right (e.g. grainsize) we should discourage brms users from using it unless they really have a problem that is computationally too demanding without it. Within chain parallelization is such an alluring feature that people are going to want to use it even when they will spend more time trying to get it right than if they just ran the model once without it. That is, there's a risk that in their effort to save time with within chain parallelization users will actually end up _wasting_ time having to try different settings, some of which make it slower than not using it at all. 

So I definitely think it's worth offering this, but I suggest advertising with caution until it basically works out of the box with not much user tweaking required. Just imagine the number of forum posts and github issues that will come in if people get really excited about this but then it requires a lot of handholding to get right.

**Comment by paul-buerkner** [comment]
I definitely agree with being cautious! Even without this feature, we are kind of getting overrun with brms questions already...

**Comment by paul-buerkner** [comment]
I have updated the branch once more to give the user control over the grainsize. Here is an example:

```R
set.seed(1234)
dat <- data.frame(
  y = rnbinom(1000, size = 10, mu = 5),
  x1 = rnorm(1000),
  x2 = rnorm(1000),
  g = factor(rep(1:100, each = 10))
)

fit1 <- brm(y ~ s(x1) + s(x2) + (1 | g), dat, family = negbinomial(), 
            chains = 1, backend = "cmdstanr", threads = NULL,
            control = list(adapt_delta = 0.95))
fit2 <- brm(y ~ s(x1) + s(x2) + (1 | g), dat, family = negbinomial(), 
            chains = 1, backend = "cmdstanr",
            threads = stanthreads(4, grainsize = 250),
            control = list(adapt_delta = 0.95))
```

In this case, the speedups are quite ok (130 vs. 67 seconds in my latest run). However, performance gains seem to be super variable, especially across likelihood family (and surely across model complexity but I haven't tested this extensively yet). For example, when changing `negbinomial` to `poisson` in the above example, I barely see any speedups. 

I really think we need more and especially systematic numerical experiments at least of the size of a student project (or did you do something like this already?). Given the relevance of within-chain parallelization for Stan and PPLs in general, this may even become a paper. No matter the details, we definitely need to learn much more about the behavior of `reduce_sum` I think.

**Comment by wds15** [comment]
I agree in that we need to setup good defaults in brms along with clear documentation. Given how brms is applying reduce_sum it seems that grainsize=1 (our current recommended starting value) is not an ideal choice for many problems. So my thinking is

- we need to define as to when this stuff gets actually useful. I see people apply paralleisation to tiny problems where its really a lot better to not bother with reduce_sum. Runtimes should be >10minutes, I think, at least.
- I would prefer if the grainsize default depends only on the data size, but not on the # of CPUs
- the grainsize should be either linear in the data set size or go with the square root of the data set size... this is my intuition which we should benchmark
- ideally the grainsize also depends on the sampling distribution, since some are far more costly to calculate (=> smaller grainsize) than others (a normal lpdf is super cheap and you need large grainsize)... but to start I would actually like to avoid this level of engineering.
- a scaling study would be good here. Hard scaling (fixed problem size, increasing CPU core count) and soft scaling (fixed work per CPU and increasing CPU count). Soft scaling lets you find out how big you can scale the problem size and still counter it with CPU power

I agree in that a full publication would be great and it‘s really worth publishing this, but we need a sutdent for that.

How about we (maybe even I) setup a small simulation study... though I am rather out of time and would welcome anyones hand here.

In any case - this is amazing to see this landing in brms!

**Comment by paul-buerkner** [comment]
Aki suggested we might want to ask on twitter or discourse if anyone is interested and has the time to work with us on the experiments.

**Comment by mike-lawrence** [comment]
This is stuff I think I can help with. I’be done lots of hierarchical multivariate simulation work. I also happened to just get access to gce’s 224 core instances, so can explore to pretty large scales. Lemme read back through this thread (only read the last two entries) and get up to speed on what parameters and scenarios you’re looking to

**Comment by paul-buerkner** [comment]
Nice! That is great1 I think we need to have a call at some point to actually define a proper scope. Right now the discussion is mostly driven from having identified we need to understand `reduce_sum` better but we didn't define what exactly we are planning to do. I will put you on my list of people to invite to the call.

**Comment by rok-cesnovar** [comment]
@mike-lawrence great!

This is a really interesting problem. It is pretty difficult to do such tuning for a general problem, but if we limit the search space to a still but less general space of brms models, I think its possible to get useful results for users and a publication.

The input (data) is: 
- problem size & complexity 
- cache size & speed
- CPU cores
- cost of parallelization (cost of spawning/joining threads + copying to the thread-local storage)

The tunable parameters are:
- grainsize
- number of threads

**Comment by wds15** [comment]
Is it for brms an option to have a routine which does some quick profiling to test on the fly for a good grainsize given some problem?

Other that that we should keep it simple to get going. It’s about getting something reasonable to work for almost any model...not the optimum.

Ideally an advanced user can always provide a little extra code to get far more optimized parallelization.

**Comment by paul-buerkner** [comment]
Can you describe a little more what you mean by quick profiling on the fly?

**Comment by rok-cesnovar** [comment]
Basically to try out a few grainsizes and pick the fastest one. This is probably a bit out of scope for brms and would rather be a part of the reduce_sum backend.

Something like: 
- try grainsize = 1
- try some other grainsizes proportional to N and sqrt(N)

Its not quite the same, but in the GPU implementation tuning research field, a quasi-random search guided with good prior knowledge outperforms any "smart" auto-tuning algorithms (based on simulate annealing or others) because of the cost of those algorithms. 

I agre that goal is always going to be good performance not optimal.

**Comment by paul-buerkner** [comment]
Thanks! That makes sense. I am not sure how I could put this into brms though. As you point out, it should probably better be part of reduce_sum itself if feasible.

**Comment by mike-lawrence** [comment]
Would it work to simply run with method="optimizing"￼ and a single iteration?￼

**Comment by rok-cesnovar** [comment]
I think that a few optimization iterations would give a good estimate yes.

**Comment by wds15** [comment]
There should not be too much variance in the measurements for problems where this actually makes sense to use. These problems are large and there should not be too much variation.

**Comment by jgabry** [comment]
> Would it work to simply run with method="optimizing"￼ and a single iteration?￼

Not with brms (doesn't support optimization), but yeah with Stan in general running a few iterations of optimization could be useful.

**Comment by paul-buerkner** [comment]
For this purpose, I could theoretically support optimization internally but not make it user facing I guess. Still not sure where this profiling thing should live ideally.

**Comment by jgabry** [comment]
I think ideally brms wouldn't have to do this and it would be done in the background by `reduce_sum()`.

**Comment by wds15** [comment]
I think a better place would be Stan services for such a profiling thing.

Anyway, I think we can make a first version available with just a reasonably large grainsize as default. This should make things work for most data sets and problems is my expectation...which should be put to a rough test before being released. This, we should aim for simple and robust as a start. I am happy to suggest a benchmark for this.

**Comment by paul-buerkner** [comment]
Some benchmarks would be really nice! Right now, when I try threading in brms with a model with some new structure I can barely predict what will happen and we definitely need to get a better understanding before we advocate or recommend anything.

**Comment by wds15** [comment]
I just noticed that you added the parallelisation to GP models... which is great to have it their, but I would be curious to see an example. In case the cholesky decomposition is outside of the reduce step, which I would expect, then not a lot of the program runs in parallel. So only when the bulk of the work of a brms program is in the reduce step we can expect good speedups.

**Comment by paul-buerkner** [comment]
Try out `gp(..., k = <number of basis functions>, gr = FALSE)` for approximate GPs without grouping. There we make full use of reduce_sum. You need to use the latest github version though as I just optimized the code yesterday.

More generally, brms now allows reduce_sum for all models that are supportable in theory, that is, all models without residual autocorrelation. For some of them, not much speed up is expected but at least it is possible to specify them that way.

**Comment by paul-buerkner** [comment]
Here is an example which could be interesting to better unstand the improvements through threading. The density of the Conway-Maxwell-Poisson (COM-Poisson) distribution is computationally very expensive and even smallish data set take ages to fit, which renders it a good canditate for threading. Unfortunately, in the one example I have run so far, speedups are quite small (around 30% speedup with 4 threads), which I find super strange as I would have expected much more speedup. However, there is also substantial variability between different runs for this family so my results may not be representative. Here is the code in case anybody is interested to run more thorough experiments for this model:

```R
data(Articles, package = "Rchoice")

# Focusing only on the students with at least one publication
phdpublish <- subset(Articles, art > 0)
phdpublish <- transform(phdpublish, art = art - 1)

# Standardise all non-binary covariates
phdpublish <- cbind(
  phdpublish[, 1:3], 
  scale(phdpublish[, -(1:3)])
)

# model without threading
fit1 <- brm(
  art ~ fem + mar + kid5 + phd + ment,
  data = phdpublish,
  family = brmsfamily("com_poisson"),
  prior= prior(normal(0, 5), class = "b"),
  chains = 1,
  backend = "cmdstanr"
)
summary(fit1)

# model with threading
fit2 <- brm(
  art ~ fem + mar + kid5 + phd + ment,
  data = phdpublish,
  family = brmsfamily("com_poisson"),
  prior= prior(normal(0, 5), class = "b"),
  chains = 1,
  backend = "cmdstanr",
  threads = threading(4, grainsize = 80)
)
summary(fit2)
```

**Comment by wds15** [comment]
I would not go straight to 4 cores, but start with 2. I am getting for a fixed seed, which you really need, these numbers:

- 1 core (no reduce sum, but compiled with threading) 160s
- 2 cores: 130s
- 3 cores: 110s

It's clearly faster, but not amazing. I don't see immediately where this is coming from... certainly this density is different than the others as it's calculated in a very basic way with lots of if's and no vectorization.

**Comment by paul-buerkner** [comment]
Thank you for checking out this example! Indeed the results don't look so amazing, although this model should be an ideal canditate. 

- The likelihood is extremely costly to evaluate.
- No vectorization is implemented and so the likelihood runs in a loop of observations anyway even if reduce_sum is not used.
- There are no big chunks of parameters being unnecessarily passed by value to each CPU. 

So I would think speed ups should be really high. @wds do you have an intuition on why this is not the case?

**Comment by wds15** [comment]
Maybe there is a function being called which triggers thread synchronization (like the lgamma did). It could be worthwhile to file an issue for this in stan-math... unless @rok-cesnovar or @bbales2 sees anything immediate.

**Comment by rok-cesnovar** [comment]
Nothing jumps out, sorry. Will try to profile a bit, definitely an interesting case.

**Comment by wds15** [comment]
@paul-buerkner What could be the case with this distribution is that it is extremely memory bound performance here. I am speculating here that what is done is to create a very large AD stack with not too complicated things to calculate for each term. Since memory access is vastly slower than CPU calculations, the parallelism does not buy too much. If this is an explanation here, then the speedups should improve on machine with larger CPU caches.

**Comment by paul-buerkner** [comment]
Thank you for the possible explanation. The more I delve into these things and ask you guys questions, the more I realize how difficult the topic actually is. Thank you for helping me on that journey. I think we should really set up a more formal study as this seems to have so many layers that we (and users) should be aware of. Should we schedule a call for this purpose? So far, I have @wds, @rok-cesnovar, @mike-lawrence on my list. Anybody I missed you may be interested in being involved in this project at this point?

**Comment by wds15** [comment]
A meeting would probably be good.. it's just that my availability is quite bad. Basically I could join this in the evenings (not this week, but next week)... let's work this out by email? BTW.. it's @wds15  for my github me.

It would help if you, @paul-buerkner, could define upfront what the goals are here from your perspective.

From my view we need

1. setup sensible defaults for most stuff in brms
2. give a sensible guidance as to when to use and when NOT to use `reduce_sum` (point to Amdahls law for expectations)
3. give a quick recipe to users (in the form of an example possibly) which lets users find out if their problem works ok with `reduce_sum` and when it does NOT work well

All of the 3 things should not be too much work to get done. This is intentionally not aiming to give users the best possible performance for their models. Instead users need to understand when to use it and when NOT to use it.

Optionally we should eventually work out a path how users can tune the performance easily (possibly with some `stancode` extra stuff), but that's probably not for now.

**Comment by paul-buerkner** [comment]
Sorry for misspelling your github name. The goals you mentioned are mostly what I had thought of as well. I will formulate them in my own words in order of my percieved importance.

1. (Related to your 2). Understand which (aspects of) models benefit from the use of `reduce_sum` and which do hinder its use. Understand the degrees to which this is the case, to give users guidance which combination of model aspects make it likely that `reduce_sum` will be beneficial from their problem (where we need to argue what kind of speedup we see as beneficial). This is of course an open ended problem so we need to restrict is somehow to certain general purpose aspects of models. Still, I think the project would strongly benefit if it tried to make claims beyond brms' implementation even if we end up using brms for most of the Stan code generation. I agree with your notion that the the results should probably be formulated in some YES not NO decision for the user to reduce complexity but still degrees of improvements beyond yes or no could be very interesting in themselves.
2. (Related to your 3). Develop an ideally automated diagnostic which can infer if `reduce_sum` is likely to be beneficial for the problem based on what we learned from 1. Since brms knows so much about its own models, we should be able to develop at least a rough automatic diagnostic for this puporse.
3. (Related to your 1). Set sensible defaults for `grainsize` and potentially other things related to `reduce_sum` when the user chooses to apply it. 

I am less optimistic than you about how much work this will require. At least, it sounds like a lot of work to me when we want to get things right in a comprehensive enough manner.

**Comment by wds15** [comment]
Hmm... I am not yet sure I understand your motivations here. So you seem to suggest that we actually need to understand deeply which models will run well and which ones will run not so well with `reduce_sum` - **without having run them**. I don't quite see the need to be able for us to tell this upfront. As you see this topic is really complicated. The basic mechanism behind this is Amdahl's law which limits us due to the fact that our program only runs partially in parallel. The crux is that we do not have any way to determine the exact fraction of code running in parallel and which runs serially in a given Stan program (though maybe we can even get to that by running with `prior_PD=FALSE/TRUE`).

In any case, from a user perspective we need to answer "Does my program benefit from `reduce_sum`?". The quickest way to find out about that is to just try it out. That's quick to do and gives you the best answer you can get, since only this is specific to your model, your data-set, your CPU, your OS, your compiler. These are way too many things for us to figure out as to how they interact and a quick test drive will tell you. So related to your point 2, I would just run a benchmark - ideally as fast as possible (based on subsets of the data).

The important thing is that we get this first test-drive to answer this question setup in a robust way. So this must not depend on a wrongly chosen `grainsize`, for example.

The other thing we have to make sure is to manage expectations from users. Doubling the speed with 3 cores is already a good thing; getting more is very much dependent on details.

So an empirical answer to the speedup question is the way to go from my view, which should be manageable.

**Comment by jgabry** [comment]
> In any case, from a user perspective we need to answer "Does my program benefit from `reduce_sum`?". The quickest way to find out about that is to just try it out.

If there's no way to tell ahead of time, then doesn't this result in a lot of users just wasting their time? Unless we think the number of cases where it's beneficial is definitely greater than the number cases it isn't, in which case we could maybe justify it by saying that on average it's worth trying. 

Or am I misunderstanding? Do you mean that they could just run their model for a few iterations to find out? If so that seems much easier to justify recommending.

**Comment by wds15** [comment]
@jgabry We need to caution users before using `reduce_sum` that it is only intended for big models. I also mean to just run the model with a few iterations with/without threading to see if they get anywhere useful (doing that with brms is simple given how easy brms generates the model code). We should probably investigate what is a sufficient iteration number for this and if we can sub-sample to smaller data-sets for this as well. In any case, I do not know what the "average" model is in brms, but I would assume that many models run by users should not run `reduce_sum`, because the data-sets are too small and it's not worth the trouble... still, for those cases where you really need `reduce_sum` and you got the resources, then it's a killer feature.

@paul-buerkner I was suggesting to use as slicing argument one of the random effects (ideally the fastest running random effect or even all random effects lumped together). I do understand why this would not be an ideal fit into your code gen pipeline, but I wanted to test this if it's worth the trouble... and it really is as it looks like. Below is a very simple example demonstrating the utility of doing so. It's rather simple minded and coded in a brute-force way, but you should get the gist of it and we can hopefully make it more generic. The non-optimized model gives me 1.07x speedup on 3 cores (basically nothing) while the optimized thing runs 2.4x faster on 3 cores. Either brms can code gen this structure directly... or we make it easy for users to supply model-specific partial sum functions.

This little example shows - unfortunately - how memory bound the performance is... at least for the normal lpdf case. Things may look different for a Poisson regression, which I would try next.

[normal_threads-opt_stan.txt](https://github.com/paul-buerkner/brms/files/5146000/normal_threads-opt_stan.txt)
[bench-normal_R.txt](https://github.com/paul-buerkner/brms/files/5146001/bench-normal_R.txt)

**Comment by jgabry** [comment]
> still, for those cases where you really need `reduce_sum` and you got the resources, then it's a killer feature.

Yeah definitely!

**Comment by paul-buerkner** [comment]
Thank you @wds15 for your comments. The points you raise are prescisely the reason why I wanted to schedule a call with you and the others who expressed interest in investigating `reduce_sum` more closely. I have not gotten a response from you, but perhaps I have sent it to the wrong email address? Would you mind checking whether you have received the email from me? Or send me an email from the address where I can reach you?

With regard to building something that predicts how well reduce_sum will work, I see that the user simply trying it out in some quick way is an easy and probably preferable way to do it. But if we want to go beyond that (independent of a particular brms user applying reduce_sum), learning more about the underlying aspects and how they play with each other would be very worthwhile, don't you think? (hence the call to clarify the sensibility and scope of this idea)

With regard to the random effects slicing. Of course, I do see your point but I am worried that the results of your simple models you try things out with won't carry over to more complex cases, e.g., with multiple grouping factors etc. So too simple examples in an (as you say) very complex environment, are perhaps not the best way to dertermine the merits of certain aspects of reduce_sum in models where it really matters. To better understand this, I think we really need something a bit more systematic, hence my idea to discuss the potential of a formal project investigating this.

**Comment by wds15** [comment]
The model with just one random effect is probably over-simplifying things in many circumstances, yes... but the example is designed to hit exactly into the memory issues we see with `reduce_sum` when using shared arguments for random effects. The problem with doing so is that for every partial sum evaluation you end up copying all random effects for each evaluation. Thus, this is very wasteful as you only need a subset of the random effects for a given partial sum. The normal likelihood is very cheap to evaluate such that memory overhead becomes quickly visible.

Now, turning to a poisson likelihood instead, then I would expect things to change. A Poisson likelihood must evaluate a log-gamma call which is very expensive such that this should offset any memory overhead you have due to wasteful copying. Running a problem with 10^4 rows and 100 groups (and a standard log-link) gives me these timings (units are seconds):

```R
> ## running with grainsize=10*N=10^5
> time1["elapsed"]
elapsed 
145.091 
> ## running with 3 cores, grainsize=200
> time2["elapsed"]
elapsed 
 57.095 
> time1["elapsed"]/time2["elapsed"]
 elapsed 
2.541221 
> ## running optimized version, grainsize=1
> time3["elapsed"]
elapsed 
 47.867 
> time1["elapsed"]/time3["elapsed"]
 elapsed 
3.031128 
> 
```

Thus, the in-efficient copying of random effects is still hurting us here, but not as much any more, since the calculation takes some time. To put this into perspective: We gain with an out-of-the-box brms program 2.5x on 3 cores. That is fantastic!

More random effects (with many levels) will increase the memory copying burden more, of course. All in all these results are very promising to me.

[bench-poisson_R.txt](https://github.com/paul-buerkner/brms/files/5152442/bench-poisson_R.txt)
[poisson_threads-opt_stan.txt](https://github.com/paul-buerkner/brms/files/5152446/poisson_threads-opt_stan.txt)

**Comment by wds15** [comment]
Thinking about the problem of copying all random effects for each partial sums, I figured that we should evaluate how increasing the number of chunks does slow down the program when run on a single core. For the Poisson model this gives expected results in that the brms auto-generated (with copying) code does significantly slow down when increasing the number of chunks:

![poisson_brms_slowdown](https://user-images.githubusercontent.com/12070126/91827389-10f7ee00-ec3f-11ea-9dc8-eda8d214f0e8.png)

The normal multi-level model behaves for 2 chunks a bit odd where it slows down a lot, but then everything is as before:

![normal_brms_slowdown](https://user-images.githubusercontent.com/12070126/91826863-6da6d900-ec3e-11ea-9dfd-b7774ba9fdbc.png)

From this perspective we should choose the grainsize around 1/10th of the number of rows (more chunks allows for more granular load balancing), but I assume that this is problem specific, but a graph like this will help choosing the grainsize a lot. This will minimize the additional overhead due to chunking when using random effects as part of the shared arguments.

To do the above plots, I used the `reduce_sum_static` to have well-defined partial sum sizes.

**Comment by paul-buerkner** [comment]
Thanks! These results are very helpful!

**Comment by jgabry** [comment]
> Thanks! These results are very helpful!

Indeed, thanks Sebastian!

**Comment by wds15** [comment]
I am glad it helps.

I was still bothered by the COM-Poisson model, which I wanted to understand better. First, providing in brms a non-normalized version of this density would make sampling this density incredibly faster (just a few seconds total sampling time)! There is not need for any parallelization in the example if you avoid calculating the normalizing constant Z.

Anyway, what turned out to make this model scale well is to rewrite the calculation of the normalizing constant. The original code calculates the normalization constant by summing terms until new terms become small enough. Moreover, the log factorial was calculated as a log sum (term by term). I now fixed the number of terms to 20 and I used the log-gamma function to calculate the factorial. Now the timings are much better (for 250 warmup / 250 iterations):

```R
> time1["elapsed"]
elapsed 
 33.532 
> ## running with 3 cores, grainsize=N/8=80
> time2["elapsed"]
elapsed 
  9.197 
> time1["elapsed"]/time2["elapsed"]
 elapsed 
3.645972 
> 
```

I think we see more than 3x speedup on 3 cores due to a better fit into the most inner CPU caches.

To me this shows that deeply nested AD tapes inside each chunk are not good for the performance. After all we do a nested AD sweep with each chunk evaluation (but I am still surprised to see such a slowdown; hopefully one can improve this in stan-math).

New log_Z function (if 20 terms are enough... I don't know!!!):
```stan
  // log normalizing constant of the COM Poisson distribution
  // implementation inspired by code of Ben Goodrich
  // Args:
  //   log_mu: log location parameter
  //   shape: positive shape parameter
  real log_Z_com_poisson(real log_mu, real nu) {
    int num_terms = 20;
    vector[num_terms] log_Z;
    //real lfac = 0;
    //real term;
    //int M;
    //real log_thres;
    if (nu == 1) {
      return exp(log_mu);
    }
    // nu == 0 or Inf will fail in this parameterization
    if (nu <= 0) {
      reject("nu must be positive");
    }
    if (nu == positive_infinity()) {
      reject("nu must be finite");
    }
    /*
    if (log_mu * nu >= log(1.5) && log_mu >= log(1.5)) {
      return log_Z_com_poisson_approx(log_mu, nu);
    }
    */
    // direct computation of the truncated series
    /*
    M = 10000;
    log_thres = log(1e-16);
    // check if the Mth term of the series is small enough
    if (nu * (M * log_mu - lgamma(M + 1)) > log_thres) {
      reject("nu is too close to zero.");
    }
    log_Z = log1p_exp(nu * log_mu);  // first 2 terms of the series
    lfac = 0;
    term = 0;
    k = 2;
    while (term > log_thres) { 
      lfac += log(k);
      term = nu * (k * log_mu - lfac);
      log_Z = log_sum_exp(log_Z, term);
      k += 1;
    }
    */
    log_Z[1] = 0.0;
    log_Z[2] = nu * log_mu;
    for(k in 2:(num_terms-1)) {
      //lfac += log(1.0*k);
      //log_Z[k+1] = nu * (k * log_mu - lfac);
      log_Z[k+1] = nu * (k * log_mu - lgamma(k+1));
    }
    return log_sum_exp(log_Z);
  }
```

**Comment by paul-buerkner** [comment]
Thank you for looking into it and for the interesting insights about the relevance of the AD tape. 

Unfortunately, we cannot just provide an unnormalized density of the COM poisson distribution and hope something correct will come out as the normalizing constant depends on the parameters. The normalizing constant is the single main problem that bugs this otherwise really nice distribution. Also, just running the first 20 terms will not work if the mean/mode of the distribution is big enough (e.g., > 20) so that a lot of terms > 20 have some relevant contributions. 

That said, I don't claim that the current COM poisson implementation is anywhere near optimal (I keep it semi-internal for a reason after all). To improve the implementation and ideally provide a mean, rather than a mode parameterization, is a whole project on its own that I would love to tackle at some point, but it may be too much to address is at the same time with `reduce_sum` (other than learning from the COM poisson distribution what is difficult for `reduce_sum`).

**Comment by wds15** [comment]
Oh dear... must have been late yesterday. Of course, normalization constants with parameters need to stay.

Anyway, I played a bit with the implementation and now things look much better to me. For 50 warmup / 50 iterations I get these timings:

```R
> ## original brms, 1 core, running with grainsize=10*N=6400
> time1["elapsed"]
elapsed 
103.215 
> ## tuned version, 1 core, running with grainsize=10*N=6400
> time1c["elapsed"]
elapsed 
 37.055 
> ## running with 3 cores, grainsize=1
> time2["elapsed"]
elapsed 
 17.903 
> time1c["elapsed"]/time2["elapsed"]
 elapsed 
2.069765 
```

So on 3 cores we get a doubling in speed - that's fair.

This is the Stan function implementing the calculation of the normalizing constant:
```stan
  // log normalizing constant of the COM Poisson distribution
  // implementation inspired by code of Ben Goodrich
  // Args:
  //   log_mu: log location parameter
  //   shape: positive shape parameter
  real log_Z_com_poisson(real log_mu, real nu) {
    real log_Z;
    //real lfac;
    //real term;
    int k = 2;
    int M = 10000;
    int converged = 0;
    int num_terms = 50;
    //real log_thres = log(1e-16); approx -36
    if (nu == 1) {
      return exp(log_mu);
    }
    // nu == 0 or Inf will fail in this parameterization
    if (nu <= 0) {
      reject("nu must be positive");
    }
    if (nu == positive_infinity()) {
      reject("nu must be finite");
    }
    if (log_mu * nu >= log(1.5) && log_mu >= log(1.5)) {
      return log_Z_com_poisson_approx(log_mu, nu);
    }
    // direct computation of the truncated series
    // check if the Mth term of the series is small enough
    if (nu * (M * log_mu - lgamma(M + 1)) > -36.0) {
      reject("nu is too close to zero.");
    }
    log_Z = log1p_exp(nu * log_mu);  // first 2 terms of the series
    //lfac = 0;
    //term = 0;
    //k = 2;
    while (converged == 0) {
      vector[num_terms+1] log_Z_terms;
      int i = 1;
      log_Z_terms[1] = log_Z;
      while(i <= num_terms) {
        log_Z_terms[i+1] = nu * (k * log_mu - lgamma(k + 1));
        k += 1;
        if ( log_Z_terms[i+1] <= -36.0 ) {
          converged = 1;
          break;
        }
        i += 1;
      }
      log_Z = log_sum_exp(log_Z_terms[1:i]);
    }
    return log_Z;
  }
```

**Comment by paul-buerkner** [comment]
Nice! Thank you so much! Can you explain what you did change except for using the `lgamma` function instead of incrementing `log` values?

**Comment by wds15** [comment]
Doing so takes advantage of the vectorized log sum exp. in the version before you were adding each term by term and now it’s in chunks of 50 terms and then we take advantage of the analytic gradients for log sum exp in vectorized form. Using 50 terms was a guess, it depends on how many terms get added up most of the time, but it seems to work just fine.

**Comment by paul-buerkner** [comment]
I have updated the COM poisson implementation in brms according to your suggestions. Thank you again :-)

**Comment by wds15** [comment]
There is an off-by-one issue which drops the last term in the series (which is anyway zero) in the function I wrote earlier, see the corrected version below.

... but now, there is a far bigger problem which we should try to catch for users: The data-set used in your COM-Poisson example **is sorted** by outcome:

```R
> phdpublish$art
  [1]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
 [26]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
 [51]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
 [76]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[101]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[126]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[151]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[176]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[201]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0
[226]  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  1  1  1
[251]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[276]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[301]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[326]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[351]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[376]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
[401]  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  2
[426]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
[451]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
[476]  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
[501]  2  2  2  2  2  2  2  2  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
[526]  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
[551]  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
[576]  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4
[601]  4  4  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  6  6  6  6  6  6
[626]  6  6  6  6  6  6  7  8  8  9 10 11 11 15 18
> 
```

thus, the computational load per partial sum is vastly different - and even worse, the partial sum computational "heavyness" is not large for the first ones and gets increasingly larger for the very few last lines. At least, this is what I think is happening. When I now do

```R
set.seed(56456)
phdpublish_unsorted  <- phdpublish[sample.int(N),]
```

then things improve more:

```R
> ## original brms, 1 core, running with grainsize=10*N=6400
> time1["elapsed"]
elapsed 
103.215 
> ## tuned version, 1 core, running with grainsize=10*N=6400
> time1c["elapsed"]
elapsed 
  66.96 
> ## running with 3 cores, grainsize=50 (static)
> time2["elapsed"]
elapsed 
  26.64 
> time1c["elapsed"]/time2["elapsed"]
 elapsed 
2.513514 
> 
```

It's no surprise that things do not work well if the computational load is not even across the data-set.

Here is the slightly improved COM-Poisson:

```stan
  // log normalizing constant of the COM Poisson distribution
  // implementation inspired by code of Ben Goodrich
  // Args:
  //   log_mu: log location parameter
  //   shape: positive shape parameter
  real log_Z_com_poisson(real log_mu, real nu) {
    real log_Z;
    //real lfac;
    //real term;
    int k = 2;
    int M = 10000;
    int num_terms = 50;
    //real log_thres = log(1e-16); approx -36
    if (nu == 1) {
      return exp(log_mu);
    }
    // nu == 0 or Inf will fail in this parameterization
    if (nu <= 0) {
      reject("nu must be positive");
    }
    if (nu == positive_infinity()) {
      reject("nu must be finite");
    }
    if (log_mu * nu >= log(1.5) && log_mu >= log(1.5)) {
      return log_Z_com_poisson_approx(log_mu, nu);
    }
    // direct computation of the truncated series
    // check if the Mth term of the series is small enough
    if (nu * (M * log_mu - lgamma(M + 1)) > -36.0) {
      reject("nu is too close to zero.");
    }
    log_Z = log1p_exp(nu * log_mu);  // first 2 terms of the series
    //lfac = 0;
    //term = 0;
    //k = 2;
    while (1) {
      vector[num_terms+1] log_Z_terms;
      int i = 1;
      log_Z_terms[1] = log_Z;
      while(i <= num_terms) {
        log_Z_terms[i+1] = nu * (k * log_mu - lgamma(k + 1));
        k += 1;
        if ( log_Z_terms[i+1] <= -36.0 ) {
          // series is converged
          return log_sum_exp(log_Z_terms[1:(i+1)]);
        }
        i += 1;
      }
      log_Z = log_sum_exp(log_Z_terms);
    }
    return negative_infinity();
  }
```

**Comment by paul-buerkner** [comment]
Thanks you for the call today. Here is a list of what needs to be done before being able to merge this feature (please make me aware of stuff I forgot):

- Add unit tests
- Choose a good default grainsize (depending on the number of observations and the number of threads)
- Allow to update `cmdstanr` models without recompilation (currently does not work as it should)
- Improve documentation
   * Add some examples illustrating the use of `reduce_sum` and how to choose the `grainsize` (Sebastian will provide examples)
   * Make clear that `reduce_sum_static` is required for exact reproducibility
   * Make sure it is clear that this feature is experimental and dampen the hopes of users


@wds15 Changing the grainsize without recompiling the model does not yet work because updating brms model with `cmdstanr` backend does not yet work (but will work hopefully before I merge this PR). So, for now, it would be ok to show the toy examples with, say, 4 or 5 different grainsizes and just recompiling the models. Would that be ok for you as well?

The grainsize can be controlled via the `threading` function as shown in the example below:

```R
# this model just serves as an illustration
# threading may not actually speed things up here
fit <- brm(count ~ zAge + zBase * Trt + (1|patient),
           data = epilepsy, family = negbinomial(),
           chains = 1, threads = threading(2, grainsize = 100),
           backend = "cmdstanr")
summary(fit)
```

**Comment by jgabry** [comment]
> Allow to update `cmdstanr` models without recompilation (currently does not work as it should)

Can you say more about this? What kind of update? Is this something we need to change in cmdstanr or is it related to how brms is interfacing with cmdstan? I can help with this if necessary.

**Comment by wds15** [comment]
@paul-buerkner I can work around the recompilation issue by either going brute-force (doing many compiles and then just take out the compilation time) or be clever about it by converting things directly to cmdstanr style (brms spits out the stan code and the stan data). The brute-force way is brms cleaner while the second approach will work with what you have. Any preference?

**Comment by paul-buerkner** [comment]
@jgabry I think it is just a brms issue.  I will let you know once I fixed the parts on the brms side. Let's see if there is still some stuff remaining afterwards.

@wds15 I prefer you using the brute-force way for now. I will convert the code to the new version once I fixed the `update` problem.

**Comment by paul-buerkner** [comment]
Updating threads and grainsize is now possible without recompilation. Here is an example:

```R
# this model just serves as an illustration
# threading may not actually speed things up here
fit1 <- brm(count ~ zAge + zBase * Trt + (1|patient),
           data = epilepsy, family = negbinomial(),
           chains = 1, threads = threading(2, grainsize = 100),
           backend = "cmdstanr")

fit2 <- update(fit1, threads = threading(2, grainsize = 50))
```

**Comment by jgabry** [comment]
Nice!

**Comment by paul-buerkner** [comment]
Unit tests are now in place as well. After adding some illustrative examples and deciding on the default grainsize, the `reduce_sum` branch is ready for review/merging.

**Comment by wds15** [comment]
Cool beans! I will try to hurry up with my bits.

**Comment by paul-buerkner** [comment]
No worries i will not be able to work much next week anyway.

wds15 <notifications@github.com> schrieb am So., 6. Sept. 2020, 20:33:

**Comment by wds15** [comment]
Below is a small code snippet which can be used by users to evaluate their models. Maybe it is possible to include such code into a brms example section?

The notebook can be rendered as spin file. This creates multiple versions of the report. Parameters which can be varied are the sample size N, the number of groups G and the likelihood (normal or poisson).

The report creates in essence two plots: One demonstrating the slowdown due to increasing the chunking. The other demonstrating at multiple grainsizes (2xdefault, default, default/2) the change in execution time. The report runs the benchmarks at 25 and 50 iterations each. The point is that the number of iterations need to be large enough for a stable estimate such that doubling the number of iterations should result in similar curves for a given problem - otherwise users need to increase these further.

The function `benchmark_threading` is relatively generic and should be useful to many other users. One point one could improve with the function is to readout ESS/s for lp__ from the cmdstan output; right now I use the total runtime as a metric.

The default grainsize of number of rows / 2 * number of physical cores seems to work ok from what I see so far.

BTW... at first I was looking at the speedup vs cores plots mainly... but then I noticed that it's a lot better to look at the absolute runtime, since the runtime at 1 core can go up with smaller grainsizes such that higher relative speedups to that get easier, but the total wallclock time is larger.

Have a look at the codes and let me know what you think. I did follow the tidyverse approaches which I hope fits into brms.

**Comment by paul-buerkner** [comment]
Thank you so much @wds15! I am still on vacation but will take a more detailed look in a week or so. With regard to making this a brms example, I prefer adding a few dependencies to the package as possible just for this example. How easy is it to strip the code of its dependency of dplyr and tidyr? The other dependencies are either part of brms already, such as ggplot2, or can easily be replaced such as the dependency on mvtnorm.

Just on minor clarification: The default rule should be #rows / (2 * #cores) right? Above, you didn't specify brackets but from the context I assume that was a typo.

**Comment by wds15** [comment]
Oh... I would have expected dplyr and tidyr to be already in your list of deps... but if not, then of course - let's take them out. This will make the code look quite a bit different, but fine for me. So I should use *apply stuff, I suppose.

I meant  #rows / (2 * #cores).. sure. cores is the number of physical cores available. Maybe we should even put there

max(100, #rows / (2 * #cores)) 

to always have at leas a grainsize of 100.

**Comment by paul-buerkner** [comment]
Thanks for the clarifications and for offering to adjust the code. I agree
tidyverse makes it more pretty but for an example in the brms doc it may
have just too many new dependencies.

wds15 <notifications@github.com> schrieb am Sa., 12. Sept. 2020, 19:54:

**Comment by wds15** [comment]
Here is a version which does not require anything else but just `ggplot2`. It's a bit more verbose to code up. Maybe this is better suited for a `demo` as part of `brms`? Just a thought.

[evaluate-tuning-base-20200914.zip](https://github.com/paul-buerkner/brms/files/5218208/evaluate-tuning-base-20200914.zip)

**Comment by paul-buerkner** [comment]
Thank you! It looks this should almost be a small vignette for brms, not an example or demo. Also, at the moment, it does not seem to be self-contained as it uses a "params" variable that seems to come out of your folder structure(?). What I am wondering is if we need this detailed vignette/demo already now or if it makes more sense to add it later once threading in brms is less of an experimental thing. If we decide to add it later, the reduce_sum feature would be ready to merge now as everything else in in place.

**Comment by wds15** [comment]
"params" is set to some default values when you run this script without anything. So it should run out of the box.

I am inclined to say... "release right away" as experimental, but then I fear we are overwhelmed with Qs from users.

How about I turn this into a mini-vignette for a still experimental feature "reduce_sum"? Some doc in form of a vignette should be a lot better than none. For the sake of simplicity I would reduce to just one likelihood presumably.

So unless you want to release right away a new brms, I can make the next days an attempt to prepare such a mini vignette.

I am definitely fine with releasing without this, but most material is already there such that it's not too much effort now to have a mini vignette (unless I overlook something here).

**Comment by paul-buerkner** [comment]
I would prefer releasing with vignette. Thank you!

wds15 <notifications@github.com> schrieb am So., 20. Sept. 2020, 15:52:

**Comment by wds15** [comment]
Great. Let's do that... it's probably easier if I fork brms and then make a PR against your repo (or you grant me directly access to this repo...whatever you prefer).

**Comment by paul-buerkner** [comment]
If you fork the repo and then work from the reduce_sum branch that would be
perfect! Brms has a couple of other rmd vignettes from which you can see
the required header structure.

wds15 <notifications@github.com> schrieb am So., 20. Sept. 2020, 16:09:

**Comment by wds15** [comment]
I started to write this up and I am starting to populate it with text. If you want to have a look at the flow of the document, have a look here:

https://github.com/paul-buerkner/brms/compare/reduce_sum...wds15:reduce_sum

If you have any comments already now, let me know.

I will need to find a way to provide the code I wrote, but not necessarily put all of that into the document as it is rendered. So maybe I will pull out some of the utility functions and make them "sourcable" such that users can grab the easily - let's see.

**Comment by paul-buerkner** [comment]
Thanks! I think the exiting text already looks quite good! I am not sure what a good approach is too sourcing code in vignettes to be honest. Personally, I would be fine with showing large chunks of code in the vignette (for users to copy), but I understand it kind of breaks the flow a little. @jgabry and @mjskay do you have experience or suggestions with handling lots of code in vignettes?

**Comment by wds15** [comment]
Maybe I found a good solution. One can run code chunk blocks in the header of the document, but not include its output at all. When you name these code chunks it is possible with knitr to print them later on in an "Appendix" section without executing them a second time. So this allows me to avoid distracting the reading flow and still include the code in completion. Sound good?

**Comment by paul-buerkner** [comment]
Sounds good!

wds15 <notifications@github.com> schrieb am Di., 22. Sept. 2020, 18:36:

**Comment by wds15** [comment]
It's progressing nicely:

[brms_threading.zip](https://github.com/paul-buerkner/brms/files/5268490/brms_threading.zip)

I will probably drop the normal model for the sake of simplicity.

Hopefully tomorrow I have time to finish the text in a first version.

**Comment by wds15** [comment]
I just pushed a complete first version. How to proceed?

**Comment by paul-buerkner** [comment]
Nice! Can you make a PR towards brms/reduce_sum?

wds15 <notifications@github.com> schrieb am Do., 24. Sept. 2020, 22:45:

**Comment by wds15** [comment]
sure... I added two more bits and you got your PR. I should stop now as the document got almost lenghty... but now people without time can grasp the most important stuff on the first page and others can get some more details by going through the entire text.

**Comment by paul-buerkner** [comment]
Thank you so much! I will read through it later on and then merge it into `reduce_sum`.

**Comment by paul-buerkner** [comment]
Closed via #1004
